### ISSUE TITLE
Implement tray item, like in Pop Shell

### ISSUE THREAD

**Issue by gikari** [gikari/bismuth#23]
Tray should have the following features:

- [ ] Quickly enabling/disabling tiling
- [ ] Layout change
- [ ] Keyboard shortcuts dialog

**Comment by alxlg** [comment]
If possible (because opening a widget would make the window lose focus) I think it would be very useful a "toggle float for current window".

Ideally this button should be in the window titlebar but KWin has not an API for that :-/

**Comment by gikari** [comment]
Plasma Applet implementation is blocked by this: https://bugs.kde.org/show_bug.cgi?id=445129

**Comment by verahawk** [comment]
FYI, I have already started working on the applet's frontend:

<img src="https://user-images.githubusercontent.com/28950897/141836448-fa1b8064-9ffe-413a-8039-d9a3ba4675a5.png" width="370" />
<img src="https://user-images.githubusercontent.com/28950897/141836457-dc6007e3-c35b-4aab-899c-95a93f406a98.png" width="370" />
<img src="https://user-images.githubusercontent.com/28950897/141836452-ab27c5bc-fac6-4f6f-897b-e8e516023c98.png" width="370" />

**Comment by gikari** [comment]
@verahawk very nice! I actually have an idea on how to forward the requests to the KWin Script part. The QML C++ for DBus interaction plugin must be written and installed. It won't be universal for all KWin Scripts of course, like in the above bug report, but at least there is a solution.

One problem is, though, I am not sure if I have enough time right now to make it, so I may slow you down a bit. Although, we could have a frontend review. Something minimal should be OK for the first iteration.

**Comment by verahawk** [comment]
No worries, I don't have too much time on my hands now either. So, no need to rush things.

**Comment by gikari** [comment]
@verahawk By the way, I've seen your screenshot in KDE VDG Room and I think, that the layout should be a little different. As we cannot change the width of the pop-up, we might put here a list of a layouts for example or something, that takes the whole width.

![IMG_20211118_005048_457.jpg](https://user-images.githubusercontent.com/14205339/142288161-02904215-54cd-451f-8e02-5d1f0151075f.jpg)

As for the check boxes, I think that applet should only include one, that the users want to change quickly and often, such as toggle tiling. Other ones belong to the settings module. The few, that remain could go to the title area (see other plasma applets for examples)

**Comment by gikari** [comment]
Here is an example of checkboxes in a networks' applet.

![Screenshot_20211118_015932](https://user-images.githubusercontent.com/14205339/142295982-18a1c16a-4c37-4097-a24c-66b48c8bad20.png)

Also, we could use a space in the applet for the big layout overview. Where is Master located and how many windows it contains. How it is rotated etc. But this might be a bit harder to make for now.

**Comment by verahawk** [comment]
Yeah, I have similar idea about live overview that would be generated on the fly using QML rectangles.

**Comment by alxlg** [comment]
I think icons like these are really good:
![immagine](https://user-images.githubusercontent.com/26327475/143465855-21bbdf3e-8470-47b6-8d84-dbf8de320356.png)

but I think they would be better as big tiles like these ones:

![immagine](https://user-images.githubusercontent.com/26327475/143465714-ff238b6b-741f-4b3a-af4a-16e23443899c.png)

and "Enable tiling" checkbox in the header like other plasmoids

**Comment by gikari** [comment]
> but I think they would be better as big tiles like these ones:

Do you mean, that instead of the list, we could present layouts in a grid?

**Comment by alxlg** [comment]
> Do you mean, that instead of the list, we could present layouts in a grid?

Yes because bigger icons are better, otherwise they are hard to recognize.

-----

By the way for anyone looking for a workaround:

1. Right click on desktop or Dolphin > create new link to app
2. As exec command use one of the following

Toggle tiling for all windows:
`qdbus org.kde.kglobalaccel /component/kwin org.kde.kglobalaccel.Component.invokeShortcut "bismuth_toggle_float_layout"`

Toggle tiling for current window:
`qdbus org.kde.kglobalaccel /component/kwin org.kde.kglobalaccel.Component.invokeShortcut "bismuth_toggle_window_floating"`

Other shortcuts name can be found by typing in a terminal:
`qdbus org.kde.kglobalaccel /component/kwin org.kde.kglobalaccel.Component.shortcutNames`

3. Drag and drop the launcher to a Plasma panel

**Comment by verahawk** [comment]
![grid1](https://user-images.githubusercontent.com/28950897/143583329-4103df19-c47d-497b-8a31-6cb808913325.png)

![grid2](https://user-images.githubusercontent.com/28950897/143583342-5c5999ff-1b23-474b-b4da-6b6ca3d952b1.png)

This on uses `Kirigami.Card` for the cards, however I am not sure if that's the best component to use. For example, there is no active or hover style available. We would have to create those manually. Maybe whatever is used for the tray pop-up list, would be better?

![tray](https://user-images.githubusercontent.com/28950897/143584501-833570c8-f834-447c-9999-46bfe5dbbef2.png)

**Comment by gikari** [comment]
> This on uses `Kirigami.Card` for the cards, however, I am not sure if that's the best component to use. For example, there is no active or hover style available. We would have to create those manually. Maybe whatever is used for the tray pop-up list, would be better?

Given the size of cards, I think they may benefit from the button to edit a layout in the future (when layout editor is ready). I didn't see the implementation, but is it possible to make the cards use 4 columns with less spacing between them? Also, the background color seems to have too much contrast.

As for the styles - the color selector applet do have the hover outline, but I'm not sure how the selected one would look like.

In case of your last proposal, I think it's valuable too, if we put the edit action to the right click menu (is it possible?).
The highlight and selected styles could be brought from the Calendar applet.

Also, a couple of other notes:
- Three lines (hamburger) menu shouldn't have the layout selection. This duplicates functionality and have a rather large list of layouts. This menu is better suited for the right-click menu of the applet, when it's minimized. For now, I think it should be removed, but for the later we might put some functionality in there.
- In general the functionality, that is present in the mockups are sufficient for the first version: just a tiling toggle and the layout selector. I have the backend ready, but the interface between the applet and the script parts must be defined.

**Comment by verahawk** [comment]
The latest iteration:

![preview](https://user-images.githubusercontent.com/28950897/144107225-cc5cac90-b3df-4080-933e-bd9f6811f032.png)

> Also, the background color seems to have too much contrast.

Yeah, those cards stand out too much. However, I believe that the only way to change the background, would be to create our own QML component.

> if we put the edit action to the right click menu

We could just show on hover an icon-only button in the bottom corner of the card.

> Three lines (hamburger) menu shouldn't have the layout selection.

That hamburger menu is automatically created based on the context menu entries, and I am not sure if it can be removed. I do agree that it is completely unnecessary, however we can probably leave it as it is.

**Comment by alxlg** [comment]
> Yeah, those cards stand out too much. However, I believe that the only way to change the background, would be to create our own QML component.

What about using the button component like here:
![immagine](https://user-images.githubusercontent.com/26327475/144233277-c4f5dca4-9765-45cf-8c6b-5a03a0bdb041.png)
?

**Comment by gikari** [comment]
I don't see the stripes on your screenshot. What do you mean?

**Comment by verahawk** [comment]
Maybe it's my monitor, but by looking form certain angles I can see some barely noticeable stripes. It's probably the gradient, that somehow breaks with buttons of such large size. 

![0](https://user-images.githubusercontent.com/28950897/144261146-a1b17dc9-dbe5-4193-8051-a8a5d5e72716.png)

**Comment by gikari** [comment]
Now I see - the background color is indeed not plain.

**Comment by alxlg** [comment]
@verahawk can you reproduce this in other plasmoids provided by default in Plasma? If yes that must be fixed by upstream

**Comment by verahawk** [comment]
Yes, buttons from the Screen Layout applet have the same artifacts. Maybe it is related to Nvidia drivers or something? Anyway, you can change the color scheme to Breeze Dark, and check it for yourself.

To be honest, it's not that big issue, so we can use those buttons for now. I just need to fix some memory leak in the applet, and once it's done, I will create a PR.

**Comment by TCMOREIRA** [comment]
What's the current progress on this?

**Comment by verahawk** [comment]
Frontend is done, however it needs some properties exposed to bind to.

![applet](https://user-images.githubusercontent.com/28950897/169246408-bf87e53f-9ef5-4e02-87e9-6015f8bd4a33.png)

**Comment by krshrimali** [comment]
> Frontend is done, however it needs some properties exposed to bind to.
> 
> ![applet](https://user-images.githubusercontent.com/28950897/169246408-bf87e53f-9ef5-4e02-87e9-6015f8bd4a33.png)

Hey, @verahawk - Just curious if this is still being looked at? Also, will keyboard shortcuts be supported for this? Like toggling the tiling mode.

**Comment by verahawk** [comment]
> Hey, @verahawk - Just curious if this is still being looked at?

Yes, however as mentioned earlier, a work needs to be done in #335, and sadly I have no sufficient knowledge in that matter.

> Also, will keyboard shortcuts be supported for this? Like toggling the "Enable Tiling Mode" option.

All settings controlled by the applet can be changed directly using shortcuts already provided by Bismuth. Also, there is a plan to add a separate tab which will contain all Bismuth shortcuts, so users can quickly look them up without diving into settings. I haven't tackled with it yet, though.

**Comment by krshrimali** [comment]
> > Hey, @verahawk - Just curious if this is still being looked at?
> 
> Yes, however as mentioned earlier, a work needs to be done in #335, and sadly I have no sufficient knowledge in that matter.
> 
> > Also, will keyboard shortcuts be supported for this? Like toggling the "Enable Tiling Mode" option.
> 
> All settings controlled by the applet can be changed directly using shortcuts already provided by Bismuth. Also, there is a plan to add a separate tab which will contain all Bismuth shortcuts, so users can quickly look them up without diving into settings. I haven't tackled with it yet, though.

Thank you, @verahawk - this makes sense. I'll try to port some if possible, and hopefully we can come back to this once #335 is done! Appreciate your response.
